test/test_um2netcdf.py

```diff
@@ -5,7 +5,209 @@
+        with mock.patch("umpost.um2netcdf.process_mule_vars") as m_mule_vars:
+            m_mule_vars.return_value = mule_vars
+
+            with mock.patch("iris.load") as m_iris_load:
+                cubes = [air_temp_cube, precipitation_flux_cube]
+
+                for c in cubes:
+                    c.attributes = {um2nc.STASH: DummyStash(*split_item_code(c.item_code))}
+                    c.cell_methods = []
+
+                    # TODO: replace with DummyCubeWithCoords ()?
+                    #       or add coord fixtures?
+                    c.coord["latitude"] = 0.0  # FIXME
+                    c.coord["longitude"] = 0.0  # FIXME
+
+                m_iris_load.return_value = cubes
+
+                with mock.patch("iris.fileformats.netcdf.Saver") as m_saver:  # prevent I/O
+                    # mock `sman` var to prevent I/O
+                    m_sman = mock.Mock()
+                    m_saver().__enter__.return_value = m_sman
+
+                    # TODO: fix lat/lon & levels requires c.coord attribute
+                    #       use fixtures to add attrs & remove the patches?
+                    with mock.patch("umpost.um2netcdf.fix_latlon_coord") as m_coord:
+                        with mock.patch("umpost.um2netcdf.fix_level_coord") as m_level:
+                            with mock.patch("umpost.um2netcdf.apply_mask") as m_apply_mask:
+                                with mock.patch("umpost.um2netcdf.cubewrite") as m_cubewrite:
+                                    infile = "/tmp/fake_input_fields_file"
+                                    outfile = "/tmp/fake_input_fields_file.nc"
+
+                                    std_args.verbose = True  # test some warning branches
+                                    um2nc.process(infile, outfile, std_args)
+
+                                    assert m_sman.update_global_attributes.called
+                                    assert m_saver.write.called is False  # write I/O prevented
+                                    assert m_coord.called
+                                    assert m_level.called
+                                    assert m_apply_mask.called is False
+                                    assert m_cubewrite.called  # real cubewrite() should be prevented
+                                    assert m_cubewrite.call_count == 1
+                                    assert m_cubewrite.call_args_list[0].args[0] == precipitation_flux_cube
+
+
+def test_process_all_cubes_filtered(air_temp_cube, mule_vars, std_args):
+    """Ensure process() exists early if all cubes are removed in filtering."""
+    with mock.patch("mule.load_umfile"):  # ignore m_load_umfile as process_mule_vars is mocked
+        with mock.patch("umpost.um2netcdf.process_mule_vars") as m_mule_vars:
+            m_mule_vars.return_value = mule_vars
+
+            with mock.patch("iris.load") as m_iris_load:
```

Comment:
In fact can a lot of these contexts be used as a multiple context `with`? Keeps the nesting down...assuming the `m_mule_vars.return_value` doesn't need to be set early.

Reply:
Yep, that's fixed with https://github.com/ACCESS-NRI/um2nc-standalone/pull/58/commits/ec45532cc02ae1694e98804994a03a8d76e770a9, with other commits fixing the other test nesting.

This is a good step forward as future mock removal is a matter of removing a line of code. **Without** this, removing mocks requires adjusting indentation for the remaining test function body...